### ISSUE TITLE
Manual intervention on tool calls

### ISSUE THREAD

**Issue by vikigenius** [pydantic/pydantic-ai#195]
It seems like for all agents, tool calls are automatic. I.e Pydantic automatically runs the tool and passes the result back and continues the run.

Can I manually intervene such that I want to manually run the tools and manage if I want to continue the run or not depending on the result?

**Comment by samuelcolvin** [comment]
This would be solved by #142.

@dmontagu this is an argument to keep the `ctx.end_run(result)` idea.

**Comment by vikigenius** [comment]
@samuelcolvin  I am not sure if I understand the PR. Is it ending the run after the tool execution or before?

For example, if I don't use any framework in OpenAI I have to manually inspect the tool_calls and call them myself. I want to be able to do that here for some agents.

**Comment by Getphyz** [comment]
Hey

I'm seeking a method to control the sequence and manage dependencies between tool executions within Pydantic AI agents, as the current automatic execution model lacks the necessary flexibility.

For example, in a trip planner, obtaining the hotel's address is essential before booking a flight. By defining such dependencies, I can prevent the LLM from getting stuck in loops.

I'm looking forward to test Pydantic Graph and plan to implement a custom "identify_tasks" node at the beginning of the graph to list all necessary tool calls based on the user query.

In summary, I'm looking for a feature akin to bind_tools that returns a list of tool calls with their arguments, enabling manual management of execution sequences and dependencies. If there's an alternative solution to address this need, please let me know.

Thx

**Comment by rubell** [comment]
It is not exactly what you asked, but wouldn't 'agent.iter' work in this case?

```python
from pydantic_ai import CallToolsNode

agent = Agent( ...)
async with agent.iter("Some prompt.") as aiter:
    async for response in aiter:
        if isinstance(response, CallToolsNode):
            if should_break(response):
                break
```

**Comment by Kludex** [comment]
> It is not exactly what you asked, but wouldn't 'agent.iter' work in this case?
> 
> ```python
> from pydantic_ai import CallToolsNode
> 
> agent = Agent(...)
> async with agent.iter("Some prompt.") as aiter:
>     async for response in aiter:
>         if isinstance(response, CallToolsNode):
>             if should_break(response):
>                 break
> ```

Yeah, I think so. The issue is older than the `iter` introduction.

**Comment by dinhngoc267** [comment]
> > It is not exactly what you asked, but wouldn't 'agent.iter' work in this case?
> > from pydantic_ai import CallToolsNode
> > 
> > agent = Agent(...)
> > async with agent.iter("Some prompt.") as aiter:
> >     async for response in aiter:
> >         if isinstance(response, CallToolsNode):
> >             if should_break(response):
> >                 break
> 
> Yeah, I think so. The issue is older than the `iter` introduction.

Hi @Kludex but there is not instruction on how to construct the messages manually through the `iter` :( 

I have to do something like this: 

```
      async with feedback_agent.iter("",
                                      deps=ctx.deps,
                                      message_history=ctx.state.message_history,
                                      model_settings={"temperature": 0.0, "parallel_tool_calls": False}) as run:

            node = run.next_node
            while not isinstance(node, End_):
                if hasattr(node, 'request'):
                    new_messages.append(node.request)
                elif hasattr(node, 'model_response'):
                    new_messages.append(node.model_response)

                if Agent_.is_call_tools_node(node):
                    part = node.model_response.parts[0]
                    tool_name = part.tool_name
                    args = json.loads(part.args)
                    tool_call_id = part.tool_call_id

                    if "final_result" in tool_name:
                            model_request = ModelRequest(parts=[
                                ToolReturnPart(
                                    tool_name=tool_name,
                                    content="Final result processed.",
                                    tool_call_id=tool_call_id,
                                    part_kind='tool-return'
                                )
                            ])
                            new_messages.append(model_request)
                            result = End(data=FeedbackOutput(**args))
                            break

                node = await run.next(node)
                if isinstance(node, End_):
                    result = node.data
                    new_messages = run.result.new_messages()
```

Can you check on the feature that we can call `run.result.new_messages()` even when run is not `End`?

**Comment by Whadup** [comment]
> > It is not exactly what you asked, but wouldn't 'agent.iter' work in this case?
> > from pydantic_ai import CallToolsNode
> > 
> > agent = Agent(...)
> > async with agent.iter("Some prompt.") as aiter:
> >     async for response in aiter:
> >         if isinstance(response, CallToolsNode):
> >             if should_break(response):
> >                 break
> 
> Yeah, I think so. The issue is older than the `iter` introduction.

Is there an elegant way to resume this in a second iter() call. When I provide iter(message_history=) with a message history that ends in the ToolCall, this tool call is not executed as expected. But doing it manually and also insertig a ToolReturnPart like @dinhngoc267 proposes seems weird to me

**Comment by jonhilgart22** [comment]
Following

My use case is the ability to decide when I need to have 'intelligence' to pick specific tools vs just running a DAG of tools without the agent picking.

**Comment by Whadup** [comment]
> My use case is the ability to decide when I need to have 'intelligence' to pick specific tools vs just running a DAG of tools without the agent picking.

The current version of pydantic_ai allows to start `iter` with a message history with open ToolCallParts which are executed first.

**Comment by DouweM** [comment]
I believe this has been addressed. If not, please file a new issue.
